### PR TITLE
zigzag: cache optimizations

### PR DESCRIPTION
**Pull request by schomatis**

### CODE REVIEW

**Comment by schomatis** [comment]
If I understand correctly, the idea is to hold off on merging this until a configuration API is in place (which I think is captured in https://github.com/filecoin-project/rust-fil-proofs/issues/501) that would help adjust the knobs of this cache.

**Comment by porcuquine** [comment]
That is correct. Please coordinate with @sidke about ETA on that feature.

**Comment by schomatis** [comment]
@porcuquine rebased and unlimited, go wild :running_man:

**Comment by porcuquine** [comment]
Thank you.

**Comment by schomatis** [comment]
@sidke delivered so it's my turn to push this forward (next week).

**Comment by schomatis** [comment]
@porcuquine heads-up, this PR will be changing in the following days (so make sure to `cherry-pick` what you need before that).

**Comment by schomatis** [comment]
Depends on (and will adapt to) https://github.com/filecoin-project/rust-fil-proofs/pull/539.

**Comment by porcuquine** [comment]
@schomatis Now that #539 has merged, I think you are clearly to lightly adapt and finally get this one merged. Thank you for your patience.

**Comment by schomatis** [comment]
@porcuquine Adapted to the new config, ready for review.

**Comment by schomatis** [comment]
(Rebasing.)

File: storage-proofs/src/zigzag_graph.rs [hunk 1]
```diff
@@ -42,6 +71,20 @@ where
...
+                    nodes
+                } else {
+                    0
+                }
+            }
+            Result::Err(_) => 0,
+            // If we can't find the `MAXIMIZE_CACHING` assume the conservative
+            // option of no cache.
+        };
+
```
Comment (porcuquine) [inline storage-proofs/src/zigzag_graph.rs]: This looks good for now. We will probably move this logic into another layer later, but putting it at the point of use seems optimal for present purposes.

**Pull request merged**